src/main/java/org/openrewrite/micrometer/TimerToObservation.java

```diff
@@ -0,0 +1,144 @@
+                    private final ChangeMethodName changeRecord = new ChangeMethodName(OBSERVATION + " record*(..)", "observe", null, null);
+
+                    @Override
+                    public J.CompilationUnit visitCompilationUnit(J.CompilationUnit compilationUnit, ExecutionContext executionContext) {
+                        J.CompilationUnit cu = compilationUnit;
+                        cu = (J.CompilationUnit) changeTypeRegistry.getVisitor().visit(cu, executionContext);
+                        cu = (J.CompilationUnit) changeTypeTimer.getVisitor().visit(cu, executionContext);
+                        cu = (J.CompilationUnit) changeRecord.getVisitor().visit(cu, executionContext);
+                        assert cu != null;
+                        return super.visitCompilationUnit(cu, executionContext);
+                    }
+
+                    @Override
+                    public J.MethodInvocation visitMethodInvocation(J.MethodInvocation mi, ExecutionContext executionContext) {
+                        if (registerMatcher.matches(mi)) {
+                            Expression timerName = null;
+                            Expression registry = mi.getArguments().get(0);
+
+                            List<String> builder = new ArrayList<>();
+                            List<Expression> parameters = new ArrayList<>();
+
+                            Expression maybeBuilder = mi.getSelect();
+                            while (maybeBuilder instanceof J.MethodInvocation) {
+                                J.MethodInvocation builderMethod = (J.MethodInvocation) maybeBuilder;
+                                if (builderMatcher.matches(maybeBuilder)) {
+                                    timerName = builderMethod.getArguments().get(0);
+                                }
+                                else if (tagMatcher.matches(maybeBuilder)) {
+                                    builder.add("\n.lowCardinalityKeyValue(#{any(String)}, #{any(String)})");
+                                    parameters.add(builderMethod.getArguments().get(0));
+                                    parameters.add(builderMethod.getArguments().get(1));
+                                }
+                                else if (tagsIterableMatcher.matches(maybeBuilder)) {
+                                    builder.add("\n.lowCardinalityKeyValues(KeyValues.of(#{any(Iterable)}, Tag::getKey, Tag::getValue))");
+                                    parameters.addAll(builderMethod.getArguments());
+                                }
+                                else if (tagsMatcher.matches(maybeBuilder)) {
+                                    String args = StringUtils.repeat("#{any(String)},", builderMethod.getArguments().size());
+                                    args = args.substring(0, args.length() - 1);
+                                    builder.add("\n.lowCardinalityKeyValues(KeyValues.of(" + args + "))");
```

Comment:
Might be unsafe to assume all of these are low cardinality. We can log and fix that in a subsequent improvement, but curious to hear your thoughts too.

Reply:
I've just decided to go with low cardinality because that's what the referenced issue was suggesting, but I agree with you that we do not have any guarantee here that this is the right choice... Are you ok going with lowCardinality for now and iterate over it later?

Reply:
If we're unsure I think high cardinality is the safer option, unless we know the assigned value is a literal or constant. That could keep things safer for now, or we risk blowing up some metrics backends.